packages/flutter_tools/lib/src/dart/pub.dart

```diff
@@ -501,7 +598,7 @@ class _DefaultPub implements Pub {
       'FLUTTER_ROOT': flutterRootOverride ?? Cache.flutterRoot!,
       _kPubEnvironmentKey: await _getPubEnvironmentValue(context),
     };
-    final String? pubCache = _getRootPubCacheIfAvailable();
+    final String? pubCache = _getLocalPubCacheIfAvailable();
     if (pubCache != null) {
```

Comment:
Can we make `pubCache` non-nullable, and always tell pub where we know the pub cache is?

Reply:
The problem comes if the localPubCache and the globalPubCache are both empty we still need to depend on pub's logic to create the global and skip setting the `PUB_CACHE` environment variable

Reply:
good point

Reply:
can we rename the function `_getPubCacheIfAvailable()` and have it return the global cache if we have one?